staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -350,8 +363,22 @@ var newETCD3Client = func(c storagebackend.TransportConfig) (*clientv3.Client, e
 		TLS:                  tlsConfig,
 		Logger:               etcd3ClientLogger,
 	}
-
-	return clientv3.New(cfg)
+	etcdClient, err := clientv3.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+	consistentListFromCacheEnabled := utilfeature.DefaultFeatureGate.Enabled(genericfeatures.ConsistentListFromCache)
+	if consistentListFromCacheEnabled {
+		if err := checkVersion(etcdClient, dialTimeout); err != nil {
+			// disable the feature
+			rerr := utilfeature.DefaultMutableFeatureGate.OverrideDefault(genericfeatures.ConsistentListFromCache, false)
```

Comment:
OK. So, the idea here is that we will also gate `WatchList` (turn it off) feature when `ConsistentListFromCache` is `disabled`. I guess this is fine as long as the incompatible etcd version is the only factor that can disable the feature. 

What if the etcd is in a good version but the `ConsistentListFromCache` is disabled (manually by a user) ? In such a scenario we will also disable the API streaming feature. Is this a realistic case ? Do we care ?

Reply:
We intend to disable just `ConsistentListFromCache`, why to mention disabling `WatchList`?

Please see the original issue https://github.com/kubernetes/kubernetes/issues/123192

Reply:
Sorry, I though this is about implementing a generic mechanism that could be used to check if etcd is in desired version.

Context: To decrease latency of the API streaming feature we would like to use `RequestWatchProgress` API call. However in order to use the `RequestWatchProgress` we need to know if the underlying data store is in correct version.

Reply:
Makes sense, however the current agreement is that `RequestWatchProgress` cannot be trusted for some versions of etcd https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md#bug-in-etcd-progress-notification.

So if you want to use it, it needs to be an optional behavior. This is not great, however if it's just for purpose of reducing latency, it might be good enough.

Reply:
Unfortunately, "live"/"client-go" clients are sensitive to latency to an extent that we have decided to disable the API streaming when we cannot use `RequestWatchProgress`.

Would you be open to creating a generic mechanism that could be used either to detect whether it is safe to use `RequestWatchProgress` or for checking the etcd version? Having a generic mechanism would allow other features to determine what kind of functionality they can offer.

Reply:
Sounds good, @ah8ad3 could you look into that? Instead of trying to disable the feature flag directly, we should add a method to `storage.Interface` that could be used to check if `RequestWatchProgress` can be used safely. This variable could be in places we already check `ConsistentListFromCache` feature flag. This would allow @p0lyn0mial to use it also for `WatchList`.

Reply:
Sounds good, let me take a look at that

Reply:
Thank you! Feel free to experiment with the design. I can also imagine having a generic mechanism that accepts a list of key-value pairs, where the key corresponds to the feature gate, and the value corresponds to the minimal etcd version. The generic mechanism could disable the feature gate if the etcd version doesn't match.

Additionally, the generic mechanism could be used during server startup to determine whether we should crash the server. For instance, if the admin explicitly enabled the feature but the etcd version doesn't match.

Reply:
That would require to create a centralized place that would track all dependencies on of K8s feature gates on etcd behavior. Don't think that would be maintainable or scale. For now let's keep the logic close to feature flag implementation. 

One thing we should be more clear with the check. Instead of checking for deprecated version, we should check for `RequestWatchProgress` working correctly. We should consider returning an error from  `RequestWatchProgress` if etcd version is too old and adding a function `CanRequestWatchProgress() bool` that could be used to decide if to use it.

Reply:
From my perspective, as soon as the streaming is requested, I would like to know whether etcd will support it (`RequestWatchProgress`). Note that this step might not even involve actually calling `RequestWatchProgress` (I think that your case will be similar).

I like the idea of disabling the feature gates when the version doesn't match. However, disabling the feature gates at runtime presents some challenges. Perhaps we should update the description of the feature, stating that it depends on the etcd version and might be disabled dynamically (at runtime). For components that expose feature gates via metrics (kube-apiserve does that), we should ensure that the actual state will be recorded (I think that today metrics are registered only once - during server startup).

Reply:
Good point about metrics, looks disabling feature gates at runtime might require more work than anticipated. Would be good to get opinion from owners of feature gate library.  cc @kubernetes/api-reviewers @kubernetes/api-approvers

Reply:
cc @deads2k

Reply:
Maybe running this line again would rewrite metrics, i can check tests of that section and maybe write test about it.
https://github.com/kubernetes/kubernetes/blob/2016fab3085562b4132e6d3774b6ded5ba9939fd/cmd/kube-apiserver/app/server.go#L117

and the test:
https://github.com/kubernetes/kubernetes/blob/2016fab3085562b4132e6d3774b6ded5ba9939fd/staging/src/k8s.io/component-base/featuregate/feature_gate_test.go#L427

I agree on asking @deads2k to find out what other affect should we consider for changing feature flag on runtime.

Reply:
Talked with @liggitt and @sftim, they don't think that feature should depend on etcd version and feature can be disabled even if feature gate is true. They proposed to use an internal bool to control whether the feature work, and not change feature gate.

Reply:
(I'm fine with emitting a metric linked to that internal bool)